infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -573,73 +563,185 @@ private LogData getLogData(LogEntry entry) {
+     * @throws IOException
+     */
+    private LogEntry parseEntry(FileChannel ch, Metadata metadata) throws IOException {
+        if (metadata == null) {
+            // The metadata for this entry was partial written
+            return null;
+        }
 
-                channelOffset += metadata.getLength();
+        ByteBuffer buffer = getPayloadForMetadata(ch, metadata);
+        if (buffer == null) {
+            // partial write on the entry
+            // rewind the channel position to point before
+            // the metadata field for this partially written payload
+            ch.position(ch.position() - METADATA_SIZE);
+            return null;
+        }
 
-            } catch (InvalidProtocolBufferException e) {
+        if (!noVerify) {
+            if (metadata.getPayloadChecksum() != getChecksum(buffer.array())) {
+                log.error("Checksum mismatch detected while trying to read file {}",
+                        ch);
                 throw new DataCorruptionException();
             }
         }
+
+
+        LogEntry entry;
+        try {
+            entry = LogEntry.parseFrom(buffer.array());
+        } catch (InvalidProtocolBufferException e) {
+            throw new DataCorruptionException();
+        }
+        return entry;
+    }
+
+    /**
+     * Reads an address space from a log file into a SegmentHandle.
+     *
+     * @param sh  Object containing state for the segment to be read
+     */
+    private void readAddressSpace(SegmentHandle sh) throws IOException {
+        FileChannel fc = sh.getWriteChannel();
+        fc.position(0);
+
+        LogHeader header = parseHeader(fc);
+        if (header == null) {
+            log.warn("Couldn't find log header for {}, creating new header.", sh.getFileName());
+            boolean verify = true;
+            if (noVerify) {
+                verify = false;
+            }
+            writeHeader(fc, VERSION, verify);
+            return;
+        }
+
+        while (fc.size() - fc.position() > 0) {
+            long channelOffset = fc.position();
+            Metadata metadata = parseMetadata(fc);
+            LogEntry entry = parseEntry(fc, metadata);
+
+            if (entry == null) {
+                // Metadata or Entry were partially written
+                log.warn("Malformed entry, metadata {} in file {}", metadata, sh.getFileName());
+
+                // Note that after rewinding the channel pointer, it is important to truncate
+                // any bytes that were written. This is required to avoid an ambigous case
+                // where a subsequent write (after a failed write) succeeds but writes less
+                // bytes than the partially written buffer. In that case, the log unit can't
+                // determine if the bytes correspund to a partially written buffer that needs
+                // to be ignored, or if the bytes corrrespond to a corrupted metadata field.
+                fc.truncate(fc.position());
```

Comment:
truncate is by size not position. Can we be off by 1 here ? also should you not use channelOffset here ?

Reply:
If you look at the parsing helper methods all of them **maintain** the **invariant** the if a field(i.e. metadata/log header/ log entry) can't be parsed, the channel point is **moved back to the last correct position**. In this case, when parse entry returns null the channel pointer i.e. position will be equal to the correct size.

One of the test cases actually tests this path and it seems like its working as expected.

Reply:
yes you are correct. I just read the documentation for the position read and write APIs and the logic you have should work.